### ISSUE TITLE
<shogun/lib/slep/slep_mc_plain_lr.h> no  such  file or directory

### ISSUE THREAD

**Issue by folivetti** [lacava/feat#263]
Whenever a try to install FEAT I get this error during compilation. Looking at Shogun repo I can't find such file, maybe it was deprecated and longer exists in current version?

**Comment by lacava** [comment]
interesting. i don't remember seeing that file. could you share your install procedure?

our [travis install script](https://github.com/lacava/feat/blob/master/ci/.travis_install.sh) might be helpful

**Comment by folivetti** [comment]
```bash
> sudo apt-get install libeigen3-dev
> conda install -c conda-forge shogun-cpp
> export SHOGUN_LIB=~/anaconda/lib
> export SHOGUN_DIR=~/anaconda/include
> ./configure
> ./install
```

This error happens because the file `src/model/shogun/MulticlassLogisticRegression.h` tries to include `shogun/lib/slep/slep_mc_plain_lr.h` that doesn't exist.

**Comment by lacava** [comment]
1) is it possible your anaconda install path is ~/anaconda3 , not ~/anaconda/?

2) something is up with your shogun install. slep_mc_plain_lr.h is in the conda shogun package in versions 6.1.4 and 6.1.3 at least. i checked for them here https://anaconda.org/conda-forge/shogun-cpp/files, you can find them in include/shogun/lib/slep of the tar files.

checking the travis build logs, shogun-cpp-6.1.3 is the version we're compiling with. 

after install it should be in your anaconda path
```
~$ find ~/anaconda3/ -name slep_mc_plain_lr.h
~/anaconda3/pkgs/shogun-cpp-6.1.3-blas_openblash9eb233f_1002/include/shogun/lib/slep/slep_mc_plain_lr.h
~/anaconda3/include/shogun/lib/slep/slep_mc_plain_lr.h
```

**Comment by folivetti** [comment]
sorry, I mistyped the path, it was `anaconda3`. In any case, there's no "slep" directory in shogun 6.1.4 (I've downloaded some of the tar files and could not find them).

I've solved the problem by creating a new environment and installing Shogun 6.1.3 with `conda install -c conda-forge shogun-cpp=6.1.3`. Everything went well but when trying to execute the wrappertest.py I get:

```
module 'pyfeat' has no attribute 'PyFeat'
```

The C++ tests work as intended!

**Comment by lacava** [comment]
> sorry, I mistyped the path, it was `anaconda3`. In any case, there's no "slep" directory in shogun 6.1.4 (I've downloaded some of the tar files and could not find them).

i see them in  linux-64/shogun-cpp-6.1.4-h9f17e87_6.tar.bz2 in /include/shogun/lib/slep/



> I've solved the problem by creating a new environment and installing Shogun 6.1.3 with `conda install -c conda-forge shogun-cpp=6.1.3`. Everything went well but when trying to execute the wrappertest.py I get:
> 
> ```
> module 'pyfeat' has no attribute 'PyFeat'
> ```

sounds like `python setup.py install` is failing. could you attach the output of ./install ?

**Comment by folivetti** [comment]
sure thing:

```
make: Entering directory '/home/olivetti/Downloads/feat/build'
make[1]: Entering directory '/home/olivetti/Downloads/feat/build'
make[2]: Entering directory '/home/olivetti/Downloads/feat/build'
make[2]: Leaving directory '/home/olivetti/Downloads/feat/build'
[ 86%] Built target feat_lib
make[2]: Entering directory '/home/olivetti/Downloads/feat/build'
make[2]: Leaving directory '/home/olivetti/Downloads/feat/build'
make[2]: Entering directory '/home/olivetti/Downloads/feat/build'
[ 87%] Built target feat
make[2]: Leaving directory '/home/olivetti/Downloads/feat/build'
[100%] Built target tests
make[1]: Leaving directory '/home/olivetti/Downloads/feat/build'
make: Leaving directory '/home/olivetti/Downloads/feat/build'
Install python wrapper? (y/n) y
installing python module...
Version error: The working tree is dirty
Unable to get version number from git tags. Not updating version.
SHOGUN_DIR= /home/olivetti/anaconda3/envs/feat/include/
package version: 0.4.post14
Warning: passing language='c++' to cythonize() is deprecated. Instead, put "# distutils: language=c++" in your .pyx or .pxd file(s)
running install
running bdist_egg
running egg_info
writing feat.egg-info/PKG-INFO
writing dependency_links to feat.egg-info/dependency_links.txt
writing requirements to feat.egg-info/requires.txt
writing top-level names to feat.egg-info/top_level.txt
reading manifest file 'feat.egg-info/SOURCES.txt'
writing manifest file 'feat.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
creating build
creating build/lib.linux-x86_64-3.8
copying feat.py -> build/lib.linux-x86_64-3.8
copying metrics.py -> build/lib.linux-x86_64-3.8
copying versionstr.py -> build/lib.linux-x86_64-3.8
running build_ext
building 'pyfeat' extension
creating build/temp.linux-x86_64-3.8
gcc -pthread -B /home/olivetti/anaconda3/envs/feat/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/eigency -I../build/ -I../src/ -I/usr/include/eigen3/ -I/home/olivetti/anaconda3/envs/feat/include/ -I/home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/numpy/core/include -I/home/olivetti/anaconda3/envs/feat/include/python3.8 -c pyfeat.cpp -o build/temp.linux-x86_64-3.8/pyfeat.o -std=c++1y -fopenmp -Wno-sign-compare -Wno-reorder
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
In file included from /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/numpy/core/include/numpy/ndarraytypes.h:1822:0,
                 from /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from pyfeat.cpp:674:
/home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
 #warning "Using deprecated NumPy API, disable it with " \
  ^~~~~~~
In file included from ../src/feat.h:15:0,
                 from pyfeat.cpp:680:
../src/init.h:40:18: warning: ‘FT::MIN_FLT’ defined but not used [-Wunused-variable]
     static float MIN_FLT = std::numeric_limits<float>::lowest();
                  ^~~~~~~
../src/init.h:39:18: warning: ‘FT::MAX_FLT’ defined but not used [-Wunused-variable]
     static float MAX_FLT = std::numeric_limits<float>::max();
                  ^~~~~~~
g++ -pthread -shared -B /home/olivetti/anaconda3/envs/feat/compiler_compat -L/home/olivetti/anaconda3/envs/feat/lib -Wl,-rpath=/home/olivetti/anaconda3/envs/feat/lib -Wl,--no-as-needed -Wl,--sysroot=/ build/temp.linux-x86_64-3.8/pyfeat.o -L/home/olivetti/anaconda3/envs/feat/lib -L/home/olivetti/Downloads/feat/build/ -Wl,-R/home/olivetti/Downloads/feat/build/ -o build/lib.linux-x86_64-3.8/pyfeat.cpython-38-x86_64-linux-gnu.so -lshogun -lfeat_lib
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/egg
copying build/lib.linux-x86_64-3.8/pyfeat.cpython-38-x86_64-linux-gnu.so -> build/bdist.linux-x86_64/egg
copying build/lib.linux-x86_64-3.8/metrics.py -> build/bdist.linux-x86_64/egg
copying build/lib.linux-x86_64-3.8/feat.py -> build/bdist.linux-x86_64/egg
copying build/lib.linux-x86_64-3.8/versionstr.py -> build/bdist.linux-x86_64/egg
byte-compiling build/bdist.linux-x86_64/egg/metrics.py to metrics.cpython-38.pyc
byte-compiling build/bdist.linux-x86_64/egg/feat.py to feat.cpython-38.pyc
byte-compiling build/bdist.linux-x86_64/egg/versionstr.py to versionstr.cpython-38.pyc
creating stub loader for pyfeat.cpython-38-x86_64-linux-gnu.so
byte-compiling build/bdist.linux-x86_64/egg/pyfeat.py to pyfeat.cpython-38.pyc
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying feat.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying feat.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying feat.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying feat.egg-info/requires.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying feat.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying feat.egg-info/zip-safe -> build/bdist.linux-x86_64/egg/EGG-INFO
writing build/bdist.linux-x86_64/egg/EGG-INFO/native_libs.txt
creating dist
creating 'dist/feat-0.4.post14-py3.8-linux-x86_64.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing feat-0.4.post14-py3.8-linux-x86_64.egg
Removing /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/feat-0.4.post14-py3.8-linux-x86_64.egg
Copying feat-0.4.post14-py3.8-linux-x86_64.egg to /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
feat 0.4.post14 is already the active version in easy-install.pth

Installed /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/feat-0.4.post14-py3.8-linux-x86_64.egg
Processing dependencies for feat==0.4.post14
Searching for pandas==1.1.0
Best match: pandas 1.1.0
Processing pandas-1.1.0-py3.8-linux-x86_64.egg
pandas 1.1.0 is already the active version in easy-install.pth

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/pandas-1.1.0-py3.8-linux-x86_64.egg
Searching for Cython==0.29.21
Best match: Cython 0.29.21
Adding Cython 0.29.21 to easy-install.pth file
Installing cygdb script to /home/olivetti/anaconda3/envs/feat/bin
Installing cython script to /home/olivetti/anaconda3/envs/feat/bin
Installing cythonize script to /home/olivetti/anaconda3/envs/feat/bin

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Searching for scikit-learn==0.23.2
Best match: scikit-learn 0.23.2
Adding scikit-learn 0.23.2 to easy-install.pth file

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Searching for scipy==1.5.2
Best match: scipy 1.5.2
Adding scipy 1.5.2 to easy-install.pth file

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Searching for numpy==1.19.1
Best match: numpy 1.19.1
Adding numpy 1.19.1 to easy-install.pth file
Installing f2py script to /home/olivetti/anaconda3/envs/feat/bin
Installing f2py3 script to /home/olivetti/anaconda3/envs/feat/bin
Installing f2py3.8 script to /home/olivetti/anaconda3/envs/feat/bin

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Searching for pytz==2020.1
Best match: pytz 2020.1
Processing pytz-2020.1-py3.8.egg
pytz 2020.1 is already the active version in easy-install.pth

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages/pytz-2020.1-py3.8.egg
Searching for python-dateutil==2.8.1
Best match: python-dateutil 2.8.1
Adding python-dateutil 2.8.1 to easy-install.pth file

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Searching for joblib==0.16.0
Best match: joblib 0.16.0
Adding joblib 0.16.0 to easy-install.pth file

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Searching for threadpoolctl==2.1.0
Best match: threadpoolctl 2.1.0
Adding threadpoolctl 2.1.0 to easy-install.pth file

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Searching for six==1.15.0
Best match: six 1.15.0
Adding six 1.15.0 to easy-install.pth file

Using /home/olivetti/anaconda3/envs/feat/lib/python3.8/site-packages
Finished processing dependencies for feat==0.4.post14
done

```

**Comment by folivetti** [comment]
btw, regarding Shogun, if you get one of those uploaded by conda-forge (they are the ones installed by conda), you can't find that required include file. Maybe this is going to be fixed in the next release.

**Comment by lacava** [comment]
hm, strange. thanks for letting me know about Shogun. 

your installation looks normal to me, i'll have to think about where that module error is coming from. do you get an error if you open python in the terminal and run `from feat import Feat` ?

**Comment by folivetti** [comment]
no error from import! A minimal example:

```python
from feat import Feat
from sklearn.datasets import load_diabetes

clf = Feat()
diabetes = load_diabetes()
X = diabetes.data
y = diabetes.target
clf.fit(X,y) # this line causes this error
```

looking at the code I'm compiling, the Feat is declared in pyfeat.cpp and pyfeact.pyx, I don't know what is causing this error.

**Comment by folivetti** [comment]
in case you're trying to replicate, here's the requirements.txt of the conda environment I used 
[requirements.txt](https://github.com/lacava/feat/files/5037670/requirements.txt)

I'm currently running Ubuntu 18.04.4 LTS. If you need any more info, let me know.

**Comment by lacava** [comment]
Ok, thanks for sharing. I’ll be able to look into it next week.

**Comment by lacava** [comment]
hi @folivetti , I am guessing this is related to using the conda environment. Are you running the minimal example from a python script or a notebook? here's my environment: [featenv.yml.txt](https://github.com/lacava/feat/files/5059183/featenv.yml.txt), I will work on replicating yours.

**Comment by folivetti** [comment]
@lacava yes, I was trying the minimal example from a python script. Also, notice that the requirements.txt I attached the other day is from a fresh install.

With a slightly modified featenv.yml (attached), the installation worked! The problem must be an incompatible library version. I'm still keeping the not working environment in case you need any other information that may help you identify the issue.

[featenv.yml.txt](https://github.com/lacava/feat/files/5064069/featenv.yml.txt)

**Comment by lacava** [comment]
after some investigating it appears this is an issue with recent versions of [setuptools](https://github.com/pypa/setuptools/issues/2246). so far I have found everything compiles fine up to setuptools 45.3. i'm still running some tests but this minimal environment works and some version of it will be added to the repo:
```
name: test
channels:
    - conda-forge
dependencies:
    - python=3.*
    - eigen=3.3.*
    - cython
    - pandas
    - scikit-learn
    - setuptools=45.*
    - shogun-cpp=6.1.3
    - wheel
```

**Comment by lacava** [comment]
@folivetti I believe this issue is addressed by PR #267. shogun 6.1.4 is included as a dependency. see the [updated install guide](https://lacava.github.io/feat/install/) and the [example conda environment](http://github.com/lacava/feat/blob/master/ci/test-environment.yml)

**Comment by hengzhe-zhang** [comment]
Hi! I still got the same error when installing the latest version of shogun 6.1.3. Could you further check this problem? Of course, it is not urgent to be solved, but it really takes me a whole week to solve this problem.
